Improvement: Implement FetchContent

ole-alb opened the pull request:
Closes #1723

**_Describe your changes here:_**

This PR changes the way how we integrate our submodules. We switch from git submodule to cmake's FetchContent. With this sc and p4est are automatically downloaded when cmake is run the first time. 

UPDATE: I also changed that creation of p4est. The release installation of p4est did not link against the release version of sc. I changed it, such that the installation of p4est in release mode links against sc in release mode. 

ToDo:
- [ ] Update installation guide 

**_All these boxes must be checked by the AUTHOR before requesting review:_**
- [x] The PR is *small enough* to be reviewed easily. If not, consider splitting up the changes in multiple PRs.
- [x] The title starts with one of the following prefixes: `Documentation:`, `Bugfix:`, `Feature:`, `Improvement:` or `Other:`.
- [x] If the PR is related to an issue, make sure to link it.
- [x] The author made sure that, as a reviewer, he/she would check all boxes below.

**_All these boxes must be checked by the REVIEWERS before merging the pull request:_**

As a reviewer please read through all the code lines and make sure that the code is *fully understood, bug free, well-documented and well-structured*.
#### General
- [ ] The reviewer executed the new code features at least once and checked the results manually.
- [ ] The code follows the [t8code coding guidelines](https://github.com/DLR-AMR/t8code/wiki/Coding-Guideline).
- [ ] New source/header files are properly added to the CMake files.
- [ ] The code is well documented. In particular, all function declarations, structs/classes and their members have a proper doxygen documentation.
- [ ] All new algorithms and data structures are sufficiently optimal in terms of memory and runtime (If this should be merged, but there is still potential for optimization, create a new issue).
#### Tests
- [ ] The code is covered in an existing or new test case using Google Test.
- [ ] The code coverage of the project (reported in the CI) should not decrease. If coverage is decreased, make sure that this is reasonable and acceptable.
- [ ] Valgrind doesn't find any bugs in the new code. [This script](https://github.com/DLR-AMR/t8code/blob/main/scripts/check_valgrind.sh) can be used to check for errors; see also this [wiki article](https://github.com/DLR-AMR/t8code/wiki/Debugging-with-valgrind).

If the Pull request introduces code that is not covered by the github action (for example coupling with a new library):
  - [ ] Should this use case be added to the github action?
  - [ ] If not, does the specific use case compile and all tests pass (check manually).
#### Scripts and Wiki
- [ ] If a new directory with source files is added, it must be covered by the `script/find_all_source_files.scp` to check the indentation of these files.
- [ ] If this PR introduces a new feature, it must be covered in an example or tutorial and a Wiki article.
#### License
- [ ] The author added a BSD statement to `doc/` (or already has one).

ole-alb committed:
update gitmodules

ole-alb committed:
update submodules

ole-alb committed:
updated CMakeLists

ole-alb committed:
remove submodule

ole-alb committed:
remove submodule

ole-alb committed:
new changes

ole-alb committed:
use fetchcontent

sandro-elsweijer committed:
reorder fetchcontent and make it verbose for downloads

sandro-elsweijer committed:
Merge branch 'main' into submodule_adjustment

holke reviewed (changes requested):
Please merge with main and see if you can get the tests running.

Davknapp committed:
Merge branch 'main' into submodule_adjustment

Davknapp committed:
Merge remote-tracking branch 'origin/main' into submodule_adjustment

Davknapp committed:
Merge remote-tracking branch 'origin/main' into submodule_adjustment

Davknapp committed:
Increase cmake version

required to use cmake to parse json-file

Davknapp committed:
Unify dependency handling

descripe dependency in a json-file, where each dependency is parsed from.
Optional provide information to a cmake-flag, if the content should not be loaded via fetchContent but using some other logic

Davknapp committed:
Don't use submodules

Davknapp committed:
Get hash from dependencies.json

Davknapp committed:
Merge remote-tracking branch 'origin/main' into submodule_adjustment

Davknapp committed:
install jq to read json file

Davknapp committed:
update installation

Davknapp committed:
udpate install for sc release and p4est

Davknapp committed:
reorder steps

Davknapp committed:
skip submodule step

Davknapp committed:
Set makros for sc depending on mpi settings

Davknapp committed:
Use sc-release for p4est in release mode

Davknapp committed:
unify names in FetchContent and SystemPackag approach

Davknapp committed:
update names in queries

Davknapp committed:
Merge remote-tracking branch 'origin/main' into submodule_adjustment

Davknapp committed:
Merge remote-tracking branch 'origin/main' into submodule_adjustment

Davknapp committed:
force new p4est and sc installation

Davknapp committed:
Move sc-MPI var setting

Davknapp committed:
Debug messages

Davknapp committed:
move mpi settings

Davknapp committed:
Move fetchContent

Davknapp committed:
ignore cache for testing

Davknapp committed:
reset cache counter

Davknapp committed:
remove debug output

Davknapp committed:
Merge branch 'main' into submodule_adjustment

holke committed:
Merge remote-tracking branch 'origin/main' into submodule_adjustment

holke commented on .github/workflows/test_preparation.yml:
```diff
@@ -66,12 +66,10 @@ jobs:
     - name: Update packages
       run: sudo apt-get update && sudo apt-get upgrade -y
       # This seems to be necessary because of the docker container
+    - name: Install jq
```
please add a comment to the file: What is jq and why do we need it here?

Davknapp replied:
Done.

holke reviewed (commented):
Thanks for this addition!

Running this out of the box worked and downloaded sc and p4est.

i then tried to run it with a local sc installation (so that it does not download sc) with

`cmake .. -DT8CODE_USE_SYSTEM_SC=/path/to/install/libsc/lib -DSC_DIR=/path/to/install/libsc/cmake`

but it still downloaded libsc and installed it again.

Did i do something wrong?

If so, the please document properly how to switch of downloading and use a locally installed sc and p4est.

Did i do it correctly?

Then why does it still download sc?

Davknapp commented:
> Thanks for this addition!
> 
> Running this out of the box worked and downloaded sc and p4est.
> 
> i then tried to run it with a local sc installation (so that it does not download sc) with
> 
> `cmake .. -DT8CODE_USE_SYSTEM_SC=/path/to/install/libsc/lib -DSC_DIR=/path/to/install/libsc/cmake`
> 
> but it still downloaded libsc and installed it again.
> 
> Did i do something wrong?
> 
> If so, the please document properly how to switch of downloading and use a locally installed sc and p4est.
> 
> Did i do it correctly?
> 
> Then why does it still download sc?

Yes, the right usage would have been:

 `cmake .. -DT8CODE_USE_SYSTEM_SC=ON -DSC_DIR=/path/to/install/libsc/cmake`
 
 Can you check if it works as expected with this command? T8CODE_USE_SYSTEM_SC is a boolean option, so only ON/OFF is accepted. The default value is OFF, therefore we download SC by default.

Davknapp committed:
Add comment about the usage of jq

Davknapp committed:
update counter

holke commented:
To what extend should "-DP4EST_USE_SYSTEM_SC" be working?

When i use
  -DT8CODE_USE_SYSTEM_SC=OFF \
  -DSC_DIR=$SC_PATH \
  -DT8CODE_USE_SYSTEM_P4EST=ON \
  -DP4EST_DIR=$P4EST_PATH \
  -DP4EST_USE_SYSTEM_SC=ON

I get:
"
CMake Warning:
  Manually-specified variables were not used by the project:

    P4EST_BUILD_TESTING
    P4EST_USE_SYSTEM_SC
"

Should that combination be working or not?

holke reviewed (changes requested):
Looks good, thanks :)

Please
- answer my comment regarding P4EST_USE_SYSTEM_SC
- Add a text to NEWS.md about these changes. Particulary important for existing users to know what changes for them.
- Add a text to the Wiki and to INSTALL about using this.

Davknapp commented:
> To what extend should "-DP4EST_USE_SYSTEM_SC" be working?
> 
> When i use -DT8CODE_USE_SYSTEM_SC=OFF -DSC_DIR=$SC_PATH -DT8CODE_USE_SYSTEM_P4EST=ON -DP4EST_DIR=$P4EST_PATH -DP4EST_USE_SYSTEM_SC=ON
> 
> I get: " CMake Warning: Manually-specified variables were not used by the project:
> 
> ```
> P4EST_BUILD_TESTING
> P4EST_USE_SYSTEM_SC
> ```
> 
> "
> 
> Should that combination be working or not?

Have you tried the same configuration on main? I have tested it and I get the same warning and I can not compile it.  I don't think that it would make sense to use possible different versions of SC for t8code and p4est. Therefore this combination should not be working.

Davknapp committed:
Add description to NEWS.md

Davknapp commented:
> Looks good, thanks :)
> 
> Please
> 
>     * answer my comment regarding P4EST_USE_SYSTEM_SC
> 
>     * Add a text to NEWS.md about these changes. Particulary important for existing users to know what changes for them.
> 
>     * Add a text to the Wiki and to INSTALL about using this.

I have addressed your comment and added a paragraph in NEWS.md 
We don't have an INSTALL file though. I can update the installation guideline in our wiki, as soon as we have merged this into main.

Davknapp committed:
Update counter

holke commented on NEWS.md:
```diff
@@ -1,3 +1,9 @@
+# Updated dependencies
+
+We have switched from a submodules approach to a FetchContent approach for our dependencies towards p4est and sc. If you don't have already installed p4est and sc, calling cmake will install the right branch from their github. If you have p4est and sc installed already and want to use that version no further code will be downloaded.
```
Please describe how users can use their existing code or where they can find that description (i.e. what you will add to the installation in the Wiki).

holke replied:
We could add a line:

"Our recommendation is to delete any existing sc/ and p4est/ folders and use the automatically downloaded FetchContent."

holke reviewed (changes requested):
Great, only a minor addition to NEWS.md left :)

Davknapp committed:
Merge remote-tracking branch 'origin/main' into submodule_adjustment

Davknapp committed:
more detailed descrition in the article.

Davknapp committed:
update cache counter to run testsuite

holke reviewed (approved):
Thanks!
